File: leetcode/triangle_path_min_sum/Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        int level = triangle.size(); 
        if (level == 1) {
            return triangle[0][0]; 
        }
        
        for (int i = level - 2; i >= 0; i--) {
            for (int j = 0; j < triangle[i].size(); j++) {
                triangle[i][j] = min(triangle[i + 1][j], triangle[i + 1][j + 1]) + triangle[i][j]; 
            }
        }
        
        return triangle[0][0]; 
    }
};

File: leetcode4/Triangle.java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int height = triangle.size(); 
        
        if (height == 0) {
            return 0; 
        }
        
        List<Integer> row = new ArrayList<Integer>(triangle.get(height - 1)); 
        
        for (int i = height - 2; i >= 0; i--) {
            List<Integer> currRow = triangle.get(i);
            for (int j = 0; j < currRow.size(); j++) {
                int min = Math.min(currRow.get(j) + row.get(j), currRow.get(j) + row.get(j + 1));
                row.set(j, min); 
            }
        }
        
        return row.get(0); 
    }
}
